### PR TITLE
scikits-odes: init at 2.4.0

### PR DESCRIPTION
**Pull request by flokli**
###### Motivation for this change
This adds `scikits.odes`, a scikit offering extra ode/dae solvers, as an extension to what is available in scipy.

It's not yet compatible with sundials 4.x, so we add `sundials_3`, too.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc @idontgetoutmuch

### CODE REVIEW

**Comment by flokli** [comment]
@GrahamcOfBorg build scikits-odes

**Review by idontgetoutmuch — APPROVED** [review]
This works for me on macOS 10.13.6 (17G65).

**Comment by flokli** [comment]
One test breaks on aarch64:

```
running install tests
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-4.6.3, py-1.7.0, pluggy-0.12.0
rootdir: /nix/store/4vkp4sslmcw3a7zia81yd8y6kdpbjkl2-python3.7-scikits.odes-2.4.0/lib/python3.7/site-packages/scikits/odes/tests
collected 81 items

test_dae.py ...F                                                         [  4%]
test_dop.py .                                                            [  6%]
test_get_info.py .....                                                   [ 12%]
test_odeint.py .....                                                     [ 18%]
test_on_funcs.py ............                                            [ 33%]
test_on_funcs_ida.py ............                                        [ 48%]
test_user_return_vals_cvode.py ........................                  [ 77%]
test_user_return_vals_ida.py ..................                          [100%]

=================================== FAILURES ===================================
______________________________ TestDae.test_lsodi ______________________________

self = <odes.tests.test_dae.TestDae testMethod=test_lsodi>

    def test_lsodi(self):
        """Check the lsodi solver"""
        for problem_cls in PROBLEMS_LSODI:
            problem = problem_cls()
>           self._do_problem(problem, 'lsodi', **problem.lsodi_pars)

test_dae.py:59:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test_dae.py:34: in _do_problem
    soln = ig.step(time, z[i], zprime[i])
../dae.py:439: in step
    return self._integrator.step(t, y_retn, yp_retn)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <scikits.odes.lsodiint.lsodi object at 0xffff7af8d3c8>, t = 400.0
y_retn = array([4.70187465e-01, 3.48124556e-06, 5.29809054e-01])
yp_retn = array([4.50537891e-01, 3.22315111e-06, 5.49458886e-01])

    def step(self, t, y_retn, yp_retn = None):
        """ See dae.DaeBase
        """
        if not self.initialized:
            raise ValueError('Method ''init_step'' has to be called prior to the'
                    'first call of ''step'' method, or after changing options')
        if t > 0.0:
            if self.tstop is None:
                itask = 1
                self.call_args[3] = 1
            else:
                itask = self.call_args[3]
                self.call_args[3] = 4
            self.y, self.yp, self.t = self.__run(self.y, self.yp, self.t, t)
            self.call_args[3] = itask
        else:
            itask = self.call_args[3]
            if self.tstop is None:
                self.call_args[3] = 2
            else:
                self.call_args[3] = 5
            self.y, self.yp, self.t  = self.__run(self.y, self.yp, self.t, -t)
            self.call_args[3] = itask
        y_retn[:] = self.y[:]
        if yp_retn is not None:
>           yp_retn[:] = self.yp[:]
E           TypeError: 'NoneType' object is not subscriptable

../lsodiint.py:376: TypeError
----------------------------- Captured stdout call -----------------------------
lsodi: excess work done on this call (check all inputs).
lsodi: present residual is [-1.17961196e-15  5.42101086e-20 -6.17733149e-12]
=============================== warnings summary ===============================
test_dae.py::TestDae::test_ida_old_api
test_dae.py::TestDae::test_ida_old_api
test_dae.py::TestDae::test_ida_old_api
  /nix/store/4vkp4sslmcw3a7zia81yd8y6kdpbjkl2-python3.7-scikits.odes-2.4.0/lib/python3.7/site-packages/scikits/odes/dae.py:389: DeprecationWarning: Old api is deprecated, move to new api
    return self._integrator.init_step(t0, y0, yp0, y_ic0_retn, yp_ic0_retn)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=============== 1 failed, 80 passed, 3 warnings in 1.89 seconds ================
builder for '/nix/store/gagbwvvk5hwfmn4cj667azympm6x5lz5-python3.7-scikits.odes-2.4.0.drv' failed with exit code 1
error: build of '/nix/store/gagbwvvk5hwfmn4cj667azympm6x5lz5-python3.7-scikits.odes-2.4.0.drv' failed
```

Any idea what could be the reason for that?

**Comment by idontgetoutmuch** [comment]
I have never used the differential algebraic part of SUNDIALS. I think the only thing one can do is report it to the maintainers of scikits-odes. One of the problems I've had in the past is that if a matrix is selected at random and is close to being singular then tests will fail. Maybe it's something like that?

**Comment by flokli** [comment]
I could reproduce it multiple times, and only on aarch64, so it might be an arch-specific issue.

**Comment by FRidh** [comment]
Please move the numpy patch out. It causes this to be a mass-rebuild.

**Comment by flokli** [comment]
> Please move the numpy patch out. It causes this to be a mass-rebuild.

I'd still like to eventually see this change land in master. What about dropping it from here, and file that change against staging?

**Comment by FRidh** [comment]
yes, open a separate PR for it

**Comment by flokli** [comment]
took the numpy patch out and rebased on latest master.

**Comment by flokli** [comment]
@GrahamcOfBorg build python3.pkgs.scikits-odes

**Comment by flokli** [comment]
I opened https://github.com/bmcage/odes/issues/101 about the `aarch64-linux` test failure.

File: pkgs/development/libraries/sundials/3.x.nix
Comment (FRidh) [inline pkgs/development/libraries/sundials/3.x.nix]: Did you ask @idontgetoutmuch ? I think you should add yourself considering this is only used in a package you're adding.

Comment (FRidh) [reply pkgs/development/libraries/sundials/3.x.nix]: Just saying because not everyone is interested in maintaining multiple versions when they're not using those versions themselves.

Comment (flokli) [reply pkgs/development/libraries/sundials/3.x.nix]: Yeah, we were both working on getting scikits-odes (and its dependencies) properly into nixpkgs, to make it more accessible, and had some off-github discussion.

Also see https://github.com/bmcage/odes/issues/98 for being able to move to the recent sundials version ;-)

Comment (idontgetoutmuch) [reply pkgs/development/libraries/sundials/3.x.nix]: I am happing to be maintainer.

**Comment by flokli** [comment]
Rebased on latest master, and set meta.platforms to `x86_64-(linux,darwin)` until https://github.com/bmcage/odes/issues/101 is fixed.

**Comment by flokli** [comment]
I managed to get sundials to build with lapackSupport, so scikit-odes now has LAPACK support, too (and we can stop skipping the corresponding tests) :tada:

This might still be broken on darwin, let's wait for CI on that.

**Comment by flokli** [comment]
@GrahamcOfBorg build python3.pkgs.scikits-odes

**Comment by idontgetoutmuch** [comment]
sundials itself seems to build with lapack - hurrah! - but when I try to use it in scikits-odes I get test failures. N.B. I have *no* overlays.

```
nix-shell -I nixpkgs=~/nixpkgs -p python37Packages.scikits-odes --run "python3"
```

```
============================= test session starts ==============================
platform darwin -- Python 3.7.4, pytest-4.6.3, py-1.7.0, pluggy-0.12.0
rootdir: /nix/store/zr1yaxwiprrpgxa3caf76iy4fzjblr26-python3.7-scikits.odes-2.4.0/lib/python3.7/site-packages/scikits/odes/tests
collected 81 items

test_dae.py ....                                                         [  4%]
test_dop.py .                                                            [  6%]
test_get_info.py .....                                                   [ 12%]
test_odeint.py ..F..                                                     [ 18%]
test_on_funcs.py ............                                            [ 33%]
test_on_funcs_ida.py ............                                        [ 48%]
test_user_return_vals_cvode.py ........................                  [ 77%]
test_user_return_vals_ida.py ..................                          [100%]

=================================== FAILURES ===================================
_________________________ test_odeint_banded_jacobian __________________________

    def test_odeint_banded_jacobian():
        # Test the use of the `Dfun`, `ml` and `mu` options of odeint default bdf
        # solver

        def func(t, y, rhs):
            rhs[:] = c.dot(y)
            return 0

        def jac(t, y, fy, J):
            J[:,:] = c
            return 0

        def bjac1_rows(t, y, fy, J):
            J[:,:] = cband1
            return 0

        def bjac2_rows(t, y, fy, J):
            # define BAND_ELEM(A,i,j) ((A->cols)[j][(i)-(j)+(A->s_mu)])
            J[:,:] = cband2
            return 0

        y0 = np.ones(4)
        t = np.array([0, 5, 10, 100])

        # Use the full Jacobian.
        sol1 = odeint(func, t, y0,
                             atol=1e-13, rtol=1e-11, max_steps=10000,
                             jacfn=jac)

        # Use the full Jacobian.
        sol2 = odeint(func, t, y0,
                             atol=1e-13, rtol=1e-11, max_steps=10000,
                             linsolver='dense')
        # Use the banded Jacobian.
        sol3 = odeint(func, t, y0,
                             atol=1e-13, rtol=1e-11, max_steps=10000,
                             jacfn=bjac1_rows, lband=2, uband=1, linsolver='band')

        # Use the banded Jacobian.
        sol4 = odeint(func, t, y0,
                             atol=1e-13, rtol=1e-11, max_steps=10000,
                             jacfn=bjac2_rows, lband=1, uband=1, linsolver='band')

        # Use the banded Jacobian.
        sol5 = odeint(func, t, y0,
                             atol=1e-13, rtol=1e-11, max_steps=10000,
                             lband=2, uband=1, linsolver='band')

        # Use the banded Jacobian.
        sol6 = odeint(func, t, y0,
                             atol=1e-13, rtol=1e-11, max_steps=10000,
                             lband=1, uband=1, linsolver='band')
        # Use the diag Jacobian.
        sol7 = odeint(func, t, y0,
                             atol=1e-13, rtol=1e-11, max_steps=10000,
                             linsolver='diag')

        #use lapack versions:
        if TEST_LAPACK:
            # Use the full Jacobian.
            sol2a = odeint(func, t, y0,
                                 atol=1e-13, rtol=1e-11, max_steps=10000,
                                 linsolver='lapackdense')
            # Use the banded Jacobian.
            sol3a = odeint(func, t, y0,
                                 atol=1e-13, rtol=1e-11, max_steps=10000,
                                 jacfn=bjac1_rows, lband=2, uband=1, linsolver='lapackband')

            # Use the banded Jacobian.
            sol4a = odeint(func, t, y0,
                                 atol=1e-13, rtol=1e-11, max_steps=10000,
                                 jacfn=bjac2_rows, lband=1, uband=1, linsolver='lapackband')

            # Use the banded Jacobian.
            sol5a = odeint(func, t, y0,
                                 atol=1e-13, rtol=1e-11, max_steps=10000,
                                 lband=2, uband=1, linsolver='lapackband')

            # Use the banded Jacobian.
            sol6a = odeint(func, t, y0,
                                 atol=1e-13, rtol=1e-11, max_steps=10000,
                                 lband=1, uband=1, linsolver='lapackband')

        #finish with some other solvers
        sol10 = odeint(func, t, y0,
                             atol=1e-13, rtol=1e-11, max_steps=10000,
                             linsolver='spgmr')
        sol11 = odeint(func, t, y0,
                             atol=1e-13, rtol=1e-11, max_steps=10000,
                             linsolver='spbcgs')
        sol12 = odeint(func, t, y0,
                             atol=1e-13, rtol=1e-11, max_steps=10000,
                             linsolver='sptfqmr')

        assert_allclose(sol1.values.y, sol2.values.y, atol=1e-12, err_msg="sol1 != sol2")
        assert_allclose(sol1.values.y, sol3.values.y, atol=1e-12, err_msg="sol1 != sol3")
        assert_allclose(sol1.values.y, sol4.values.y, atol=1e-12, err_msg="sol1 != sol4")
        assert_allclose(sol1.values.y, sol5.values.y, atol=1e-12, err_msg="sol1 != sol5")
        assert_allclose(sol1.values.y, sol6.values.y, atol=1e-12, err_msg="sol1 != sol6")
        assert_allclose(sol1.values.y, sol7.values.y, atol=1e-12, err_msg="sol1 != sol7")
        assert_allclose(sol1.values.y, sol10.values.y, atol=1e-12, err_msg="sol1 != sol10")
        assert_allclose(sol1.values.y, sol11.values.y, atol=1e-12, err_msg="sol1 != sol11")
        assert_allclose(sol1.values.y, sol12.values.y, atol=1e-12, err_msg="sol1 != sol12")

        if TEST_LAPACK:
>           assert_allclose(sol1.values.y, sol2a.values.y, atol=1e-12, err_msg="sol1 != sol2a")
E           AssertionError:
E           Not equal to tolerance rtol=1e-07, atol=1e-12
E           sol1 != sol2a
E           (shapes (4, 4), (1, 4) mismatch)
E            x: array([[1.000000e+00, 1.000000e+00, 1.000000e+00, 1.000000e+00],
E                  [3.149379e-10, 6.389274e-06, 1.199541e-04, 7.440796e-03],
E                  [3.337533e-13, 6.808093e-09, 5.058055e-07, 5.041742e-05],
E                  [3.850652e-25, 7.855459e-21, 5.895533e-19, 5.900678e-17]])
E            y: array([[1., 1., 1., 1.]])

test_odeint.py:373: AssertionError
----------------------------- Captured stderr call -----------------------------

[CVODE ERROR]  CVode
  At t = 0, the setup routine failed in an unrecoverable manner.


[CVODE ERROR]  CVode
  At t = 0, the setup routine failed in an unrecoverable manner.


[CVODE ERROR]  CVode
  At t = 0, the setup routine failed in an unrecoverable manner.


[CVODE ERROR]  CVode
  At t = 0, the setup routine failed in an unrecoverable manner.


[CVODE ERROR]  CVode
  At t = 0, the setup routine failed in an unrecoverable manner.

=============================== warnings summary ===============================
test_dae.py::TestDae::test_ida_old_api
test_dae.py::TestDae::test_ida_old_api
test_dae.py::TestDae::test_ida_old_api
  /nix/store/zr1yaxwiprrpgxa3caf76iy4fzjblr26-python3.7-scikits.odes-2.4.0/lib/python3.7/site-packages/scikits/odes/dae.py:389: DeprecationWarning: Old api is deprecated, move to new api
    return self._integrator.init_step(t0, y0, yp0, y_ic0_retn, yp_ic0_retn)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=============== 1 failed, 80 passed, 3 warnings in 1.36 seconds ================
 ** On entry to DGETRF parameter number  1 had an illegal value
 ** On entry to DGBTRF  parameter number  1 had an illegal value
 ** On entry to DGBTRF  parameter number  1 had an illegal value
 ** On entry to DGBTRF  parameter number  1 had an illegal value
 ** On entry to DGBTRF  parameter number  1 had an illegal value
builder for '/nix/store/zx1w1rm5b8flkjsjd90nzxcg1hyznjv9-python3.7-scikits.odes-2.4.0.drv' failed with exit code 1
error: build of '/nix/store/zx1w1rm5b8flkjsjd90nzxcg1hyznjv9-python3.7-scikits.odes-2.4.0.drv' failed
```

**Comment by idontgetoutmuch** [comment]
https://logs.nix.ci/?key=nixos/nixpkgs.64195&attempt_id=28e2cb65-861c-4a10-b40c-be302ba382f7

has

```
============================= test session starts ==============================
platform darwin -- Python 3.7.4, pytest-4.6.3, py-1.7.0, pluggy-0.12.0
rootdir: /nix/store/zr1yaxwiprrpgxa3caf76iy4fzjblr26-python3.7-scikits.odes-2.4.0/lib/python3.7/site-packages/scikits/odes/tests
collected 81 items
test_dae.py ....                                                         [  4%]
test_dop.py .                                                            [  6%]
test_get_info.py .....                                                   [ 12%]
building of '/nix/store/zx1w1rm5b8flkjsjd90nzxcg1hyznjv9-python3.7-scikits.odes-2.4.0.drv' timed out after 3600 seconds
error: build of '/nix/store/zx1w1rm5b8flkjsjd90nzxcg1hyznjv9-python3.7-scikits.odes-2.4.0.drv' failed
```

so I am not sure why we see "all checks have passed".

**Comment by steinitznavican** [comment]
[build-2019-07-30-Tue-09:34.log](https://github.com/NixOS/nixpkgs/files/3445891/build-2019-07-30-Tue-09.34.log)

**Comment by flokli** [comment]
Huh, that's weird - it seems all the configure-phase checks are successful, we get everything to link and run without segfaulting, but actually running it produces unsatisfying results? :confused:

**Comment by flokli** [comment]
@GrahamcOfBorg build python3.pkgs.scikits-odes

**Comment by flokli** [comment]
It seems the lapack component inside sundials is still broken (even though the cmake configure phase passes), on both MacOS and Linux.

Unfortunately, the tests on linux didn't seem to show that.

I tried running sundial's tests (`doCheck = true; checkPhase = "make test";`), unfortunately, the test binaries seem to set a non-working RUNPATH (containing $out), so all tests fail.

I built one of the examples myself manually, both without lapack and with lapack.

The version without lapack seems to work:
```
⇒  cc -I/nix/store/8rx1plwcs8nn88f58lybms98x3jvznrn-sundials-3.2.1/include /nix/store/8rx1plwcs8nn88f58lybms98x3jvznrn-sundials-3.2.1/share/examples/cvode/serial/cvRoberts_dns.c -L/nix/store/8rx1plwcs8nn88f58lybms98x3jvznrn-sundials-3.2.1/lib -lsundials_cvode                                 
⇒  ./a.out
 
3-species kinetics problem

At t = 2.6391e-01      y =  9.899653e-01    3.470564e-05    1.000000e-02
    rootsfound[] =   0   1
At t = 4.0000e-01      y =  9.851641e-01    3.386242e-05    1.480205e-02
At t = 4.0000e+00      y =  9.055097e-01    2.240338e-05    9.446793e-02
At t = 4.0000e+01      y =  7.158017e-01    9.185037e-06    2.841892e-01
At t = 4.0000e+02      y =  4.505360e-01    3.223271e-06    5.494608e-01
At t = 4.0000e+03      y =  1.832299e-01    8.944378e-07    8.167692e-01
At t = 4.0000e+04      y =  3.898902e-02    1.622006e-07    9.610108e-01
At t = 4.0000e+05      y =  4.936383e-03    1.984224e-08    9.950636e-01
At t = 4.0000e+06      y =  5.168093e-04    2.068293e-09    9.994832e-01
At t = 2.0790e+07      y =  1.000000e-04    4.000397e-10    9.999000e-01
    rootsfound[] =  -1   0
At t = 4.0000e+07      y =  5.202440e-05    2.081083e-10    9.999480e-01
At t = 4.0000e+08      y =  5.201061e-06    2.080435e-11    9.999948e-01
At t = 4.0000e+09      y =  5.258603e-07    2.103442e-12    9.999995e-01
At t = 4.0000e+10      y =  6.934511e-08    2.773804e-13    9.999999e-01

Final Statistics:
nst = 542    nfe  = 755    nsetups = 107    nfeLS = 0      nje = 11
nni = 751    ncfn = 0      netf = 22     nge = 570
```

The version with lapack (`-lsundials_sunlinsollapackdense`) fails:

```
⇒  cc -I/nix/store/8rx1plwcs8nn88f58lybms98x3jvznrn-sundials-3.2.1/include /nix/store/8rx1plwcs8nn88f58lybms98x3jvznrn-sundials-3.2.1/share/examples/cvode/serial/cvRoberts_dnsL.c -L/nix/store/8rx1plwcs8nn88f58lybms98x3jvznrn-sundials-3.2.1/lib -lsundials_cvode -lsundials_sunlinsollapackdense
⇒  ./a.out                                     
 
3-species kinetics problem

*** stack smashing detected ***: <unknown> terminated
[1]    26592 abort      ./a.out
```
That was on Linux, not MacOS.

There seems to be something broken with how the fortran code is called. Any pointers?

**Comment by steinitznavican** [comment]
This works

```
cmake -DLAPACK_ENABLE=ON \
-DLAPACK_LIBRARIES="/usr/local/opt/openblas/lib/libblas.dylib;/usr/local/opt/openblas/lib/liblapack.dylib" \
-DSUNDIALS_INDEX_TYPE=int32_t \
-DCMAKE_INSTALL_PREFIX=/Users/dominic.steinitz/MySundials/instdir \
-DEXAMPLES_INSTALL_PATH=/Users/dominic.steinitz/MySundials/instdir/examples \
-DEXAMPLES_INSTALL=ON \
-DEXAMPLES_ENABLE_C=ON \
/Users/dominic.steinitz/sundials-3.1.1
```

```
> > > > > > > -- The Fortran compiler identification is GNU 9.1.0
-- Searching for a Fortran compiler... /usr/local/bin/gfortran
-- Checking whether Fortran compiler has -isysroot
-- Checking whether Fortran compiler has -isysroot - yes
-- Checking whether Fortran compiler supports OSX deployment target flag
-- Checking whether Fortran compiler supports OSX deployment target flag - yes
-- Trying to compile and link a simple Fortran program... OK
-- Determining Fortran name-mangling scheme... OK
-- Looking for LAPACK libraries... OK
-- Checking if LAPACK works... OK
-- Added NVECTOR_SERIAL module
-- Added SUNMATRIX_DENSE module
-- Added SUNMATRIX_BAND module
-- Added SUNMATRIX_SPARSE module
-- Added SUNLINSOL_BAND module
-- Added SUNLINSOL_DENSE module
-- Added SUNLINSOL_LAPACKBAND module
-- Added SUNLINSOL_LAPACKDENSE module
-- Added SUNLINSOL_SPGMR module
-- Added SUNLINSOL_SPFGMR module
-- Added SUNLINSOL_SPBCGS module
-- Added SUNLINSOL_SPTFQMR module
-- Added SUNLINSOL_PCG module
-- Added ARKODE module
-- Added CVODE module
-- Added CVODES module
-- Added IDA module
-- Added IDAS module
-- Added KINSOL module
-- Configuring done
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   sundials_arkode_shared
   sundials_cvode_shared
   sundials_cvodes_shared
   sundials_ida_shared
   sundials_idas_shared
   sundials_kinsol_shared
   sundials_nvecserial_shared
   sundials_sunlinsolband_shared
   sundials_sunlinsoldense_shared
   sundials_sunlinsollapackband_shared
   sundials_sunlinsollapackdense_shared
   sundials_sunlinsolpcg_shared
   sundials_sunlinsolspbcgs_shared
   sundials_sunlinsolspfgmr_shared
   sundials_sunlinsolspgmr_shared
   sundials_sunlinsolsptfqmr_shared
   sundials_sunmatrixband_shared
   sundials_sunmatrixdense_shared
   sundials_sunmatrixsparse_shared

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /Users/dominic.steinitz/mySundials-3.1.1
```

**Comment by steinitznavican** [comment]
The nix derivation uses cmake to generate

```
-DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON \
-DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON \
-DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON \
-DCMAKE_BUILD_TYPE=Release \
-DCMAKE_SKIP_BUILD_RPATH=ON \
-DCMAKE_INSTALL_INCLUDEDIR=/nix/store/hq8bh2mwhm10vwr9w38n2hy3zrw1q2zz-sundials-3.2.1/include \
-DCMAKE_INSTALL_LIBDIR=/nix/store/hq8bh2mwhm10vwr9w38n2hy3zrw1q2zz-sundials-3.2.1/lib \
-DCMAKE_INSTALL_NAME_DIR=/nix/store/hq8bh2mwhm10vwr9w38n2hy3zrw1q2zz-sundials-3.2.1/lib \
-DCMAKE_POLICY_DEFAULT_CMP0025=NEW \
-DCMAKE_OSX_DEPLOYMENT_TARGET= \
-DCMAKE_OSX_SYSROOT= \
-DCMAKE_FIND_FRAMEWORK=last \
-DCMAKE_STRIP=/nix/store/p20s2ajhyhq6lbjsdyywsnvz3xf41jz8-cctools-binutils-darwin/bin/strip \
-DCMAKE_RANLIB=/nix/store/p20s2ajhyhq6lbjsdyywsnvz3xf41jz8-cctools-binutils-darwin/bin/ranlib \
-DCMAKE_AR=/nix/store/p20s2ajhyhq6lbjsdyywsnvz3xf41jz8-cctools-binutils-darwin/bin/ar \
-DCMAKE_C_COMPILER=clang \
-DCMAKE_CXX_COMPILER=clang++ \
-DCMAKE_INSTALL_PREFIX=/nix/store/hq8bh2mwhm10vwr9w38n2hy3zrw1q2zz-sundials-3.2.1 \
-DEXAMPLES_INSTALL_PATH=/nix/store/hq8bh2mwhm10vwr9w38n2hy3zrw1q2zz-sundials-3.2.1/share/examples \
-DEXAMPLES_INSTALL=ON \
-DEXAMPLES_ENABLE_C=ON \
-DSUNDIALS_INDEX_TYPE=int32_t \
-DLAPACK_ENABLE=ON \
-DLAPACK_LIBRARIES=/nix/store/qcjzcrxwynyd7xyay58r3kl6qxq6qpxw-openblas-0.3.6/lib/libopenblas.dylib;/nix/store/qcjzcrxwynyd7xyay58r3kl6qxq6qpxw-openblas-0.3.6/lib/liblapack.dylib
```

**Comment by steinitznavican** [comment]
I have this

```
  postInstall = ''
                  mkdir $out/easytofind
                  echo $(ls) > $out/easytofind/foo.txt
                  cp -r LapackTest $out/easytofind/LapackTest
                  cd $out/easytofind/LapackTest
                  make
                  ./ltest > $out/easytofind/ltest.out
                '';
```

but sadly don't get what I wanted

```
/nix/store/sy5iwww8nc4a7mdb3ls44q96i0kxy2dx-cmake-3.14.5/bin/cmake -S/private/var/folders/1g/xdjj8dy15k932b80r197qcnr0000gp/T/nix-build-sundials-3.2.1.drv-0/sundials-3.2.1/build/LapackTest -B/private/var/folders/1g/xdjj8dy15k932b80r197qcnr0000gp/T/nix-build-sundials-3.2.1.drv-0/sundials-3.2.1/build/LapackTest --check-build-system CMakeFiles/Makefile.cmake 0
-- The C compiler identification is Clang 7.1.0
-- Check for working C compiler: /nix/store/b8i7jqbkq80h6v03pnnkpihiq72iplwq-clang-wrapper-7.1.0/bin/clang
-- Check for working C compiler: /nix/store/b8i7jqbkq80h6v03pnnkpihiq72iplwq-clang-wrapper-7.1.0/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /private/var/folders/1g/xdjj8dy15k932b80r197qcnr0000gp/T/nix-build-sundials-3.2.1.drv-0/sundials-3.2.1/build/LapackTest
/nix/store/sy5iwww8nc4a7mdb3ls44q96i0kxy2dx-cmake-3.14.5/bin/cmake -E cmake_progress_start /private/var/folders/1g/xdjj8dy15k932b80r197qcnr0000gp/T/nix-build-sundials-3.2.1.drv-0/sundials-3.2.1/build/LapackTest/CMakeFiles /private/var/folders/1g/xdjj8dy15k932b80r197qcnr0000gp/T/nix-build-sundials-3.2.1.drv-0/sundials-3.2.1/build/LapackTest/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/nix/store/s41pkl0pjqxfq2hxfcmfc03205njp4sc-sundials-3.2.1/easytofind/LapackTest'
make[1]: CMakeFiles/Makefile2: No such file or directory
make[1]: *** No rule to make target 'CMakeFiles/Makefile2'.  Stop.
make[1]: Leaving directory '/nix/store/s41pkl0pjqxfq2hxfcmfc03205njp4sc-sundials-3.2.1/easytofind/LapackTest'
make: *** [Makefile:87: all] Error 2
```

**Comment by steinitznavican** [comment]
This is a bit better but still fails.

```
  postInstall = ''
                  mkdir $out/easytofind
                  echo $(ls) > $out/easytofind/foo.txt
                  cp -r . $out/easytofind
                  cd $out/easytofind/LapackTest
                  make
                  ./ltest > $out/easytofind/ltest.out
                '';
```

```
/nix/store/sy5iwww8nc4a7mdb3ls44q96i0kxy2dx-cmake-3.14.5/bin/cmake -S/private/var/folders/1g/xdjj8dy15k932b80r197qcnr0000gp/T/nix-build-sundials-3.2.1.drv-0/sundials-3.2.1/build/LapackTest -B/private/var/folders/1g/xdjj8dy15k932b80r197qcnr0000gp/T/nix-build-sundials-3.2.1.drv-0/sundials-3.2.1/build/LapackTest --check-build-system CMakeFiles/Makefile.cmake 0
-- The C compiler identification is Clang 7.1.0
-- Check for working C compiler: /nix/store/b8i7jqbkq80h6v03pnnkpihiq72iplwq-clang-wrapper-7.1.0/bin/clang
-- Check for working C compiler: /nix/store/b8i7jqbkq80h6v03pnnkpihiq72iplwq-clang-wrapper-7.1.0/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /private/var/folders/1g/xdjj8dy15k932b80r197qcnr0000gp/T/nix-build-sundials-3.2.1.drv-0/sundials-3.2.1/build/LapackTest
/nix/store/sy5iwww8nc4a7mdb3ls44q96i0kxy2dx-cmake-3.14.5/bin/cmake -E cmake_progress_start /private/var/folders/1g/xdjj8dy15k932b80r197qcnr0000gp/T/nix-build-sundials-3.2.1.drv-0/sundials-3.2.1/build/LapackTest/CMakeFiles /private/var/folders/1g/xdjj8dy15k932b80r197qcnr0000gp/T/nix-build-sundials-3.2.1.drv-0/sundials-3.2.1/build/LapackTest/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/nix/store/84h3i703qzjkf876pxvc7ny9vqcqj28i-sundials-3.2.1/easytofind/LapackTest'
make[1]: CMakeFiles/Makefile2: No such file or directory
make[1]: *** No rule to make target 'CMakeFiles/Makefile2'.  Stop.
make[1]: Leaving directory '/nix/store/84h3i703qzjkf876pxvc7ny9vqcqj28i-sundials-3.2.1/easytofind/LapackTest'
make: *** [Makefile:87: all] Error 2
```

**Comment by flokli** [comment]
CMake can be convinced to use a working RUNPATH during `testPhase`, by following https://gitlab.kitware.com/cmake/community/wikis/doc/cmake/RPATH-handling#always-full-rpath. I opened https://github.com/LLNL/sundials/pull/19 upstream.

**Comment by flokli** [comment]
liblapack needed to be built with shared library support (`BUILD_SHARED_LIBS=ON`), after that, sundials was able to use everything just fine from only `liblapack`, without `openblas` involved. I guess the previous segfaults were mostly due to us using incompatible libraries there.

**Comment by flokli** [comment]
@GrahamcOfBorg build python3.pkgs.scikits-odes

**Comment by flokli** [comment]
@GrahamcOfBorg build python2.pkgs.scikits-odes python3.pkgs.scikits-odes

**Comment by idontgetoutmuch** [comment]
Any reason this can't be merged?

**Comment by flokli** [comment]
Hm. It would have been nice if we heard back from LLNL/sundials about the patches we sent upstream, but their repo seems to be only code dumps of release snapshots so far, so I don't really see if and how patches are merged in :cry:

I did not yet include the patch from https://github.com/LLNL/sundials/pull/21 to remove test flakiness into this PR. It doesn't apply cleanly on `sundials_3` (which we need for scikits-odes until https://github.com/bmcage/odes/issues/98 is solved).

I really hope LLNL/sundials is moving development to github soon as indicated on https://github.com/LLNL/sundials/pull/8#issuecomment-458289025 - a sundials 3 branch containing a rebased patch would be really cool - nixpkgs policy is to not include patches into nixpkgs directly, but fetch them if possible…

**Comment by flokli** [comment]
I included the non-flakyness patches for sundials 4, and disabled tests for `sundials_3`, as the patches don't apply cleanly there.

I also bumped scikits-odes to `2.4.0-9-g93075ae`, containing the aarch64 and macos patches from https://github.com/bmcage/odes/pull/104 and https://github.com/bmcage/odes/pull/105.

**Comment by flokli** [comment]
@GrahamcOfBorg build python.pkgs.scikits-odes python3.pkgs.scikits-odes sundials

**Pull request merged**
